udpsession.go

```diff
@@ -0,0 +1,364 @@
+		}
+	} else {
+		us.Lock()
+		us.sessKeyToSessAddr = make(map[string]*udpSessionAddr)
+		us.udpAddrToSessAddr = make(map[string]*udpSessionAddr)
+		for _, conn := range us.tcpConns {
+			if conn != nil {
+				conn.Close()
+			}
+		}
+		us.tcpConns = make(map[string]*Conn)
+		us.Unlock()
+	}
+
+	return
+}
+
+// Get UDPAddr by net.Addr
+func (us *UdpSession) getUdpAddr(addr net.Addr) (udpAddr *net.UDPAddr, err error) {
+	if us.endpointType == EndpointDialer {
+		udpAddr = us.peer.udpAddr
+	} else {
+		if addr == nil {
+			return nil, ErrWrongAddr
+		} else {
+			usa, ok := addr.(*udpSessionAddr) // addr should be *udpSessionAddr
+			if !ok {
+				return nil, ErrWrongAddr
+			}
+			us.RLock()
+			usa, ok = us.sessKeyToSessAddr[usa.String()]
+			us.RUnlock()
+			if ok {
+				udpAddr = usa.udpAddr
+			} else {
+				return nil, ErrWrongAddr
+			}
+		}
+	}
+	return
+}
+
+// Handle session management message.
+func (us *UdpSession) handleSessionMsg(b []byte, udpAddr *net.UDPAddr) error {
+	header := b[0]
+	if header != byte(pb.HeaderType_SESSION) {
+		return ErrWrongMsgFormat
+	}
+	sessData, err := parseSessionData(b)
+	if err != nil {
+		return err
+	}
+
+	switch sessData.MsgType {
+	case pb.MsgType_REMOTEADDR:
+		remoteAddr := string(sessData.Data)
+		if !us.ts.shouldAcceptAddr(remoteAddr) {
+			return ErrRemoteAddrRejected
+		}
+
+		usa := &udpSessionAddr{remoteAddr: remoteAddr, udpAddr: udpAddr}
+		us.Lock()
+		us.udpAddrToSessAddr[udpAddr.String()] = usa
+		us.Unlock()
+
+	case pb.MsgType_SESSIONID:
+		us.RLock()
+		usa, ok := us.udpAddrToSessAddr[udpAddr.String()]
+		us.RUnlock()
+		if !ok {
+			return ErrWrongAddr
+		}
+
+		metadata := &pb.SessionMetadata{}
+		err = proto.Unmarshal(sessData.Data, metadata)
+		if err != nil {
+			return err
+		}
+		sessionID := metadata.Id
+		usa.sessionID = sessionID
+		usa.udpAddr = udpAddr
+
+		us.Lock()
+		us.sessKeyToSessAddr[usa.String()] = usa // store udpSessionAddr, key is usa.String()
+		us.udpAddrToSessAddr[udpAddr.String()] = usa
+		us.Unlock()
+
+	case pb.MsgType_CLOSE:
+		if us.endpointType == EndpointDialer {
+			us.Close()
+		} else {
+			us.oneConnClosed(udpAddr)
+		}
+
+	default:
+		return ErrWrongMsgFormat
+	}
+
+	return nil
+}
+
+// Monitor connection status by tcp connection.
+func (us *UdpSession) statusMonitor() (ch chan struct{}) {
+	if us.endpointType != EndpointDialer {
+		return
+	}
+
+	tcpConn := us.tcpConn.Conn.(*net.TCPConn)
+	err := tcpConn.SetKeepAlive(true)
+	if err != nil {
+		log.Printf("udpStatusMonitor SetKeepAlive %v", err)
+	}
+
+	ch = make(chan struct{})
+	go func() {
+		defer close(ch)
+		b := make([]byte, 1024)
+		for {
+			if us.ts.IsClosed() || us.udpConn.IsClosed() {
+				break
+			}
+			_, err := tcpConn.Read(b)
+			if err != nil {
+				log.Printf("udpStatusMonitor tcpConn.Read %v", err)
+				break
+			}
+		}
+	}()
+
+	return
+}
+
+func (us *UdpSession) sendCloseMsg(udpAddr *net.UDPAddr) (err error) {
+	b, _ := newSessionData(pb.MsgType_CLOSE, nil)

```

Comment:
It's better to send close with TCP because it's more reliable

Reply:
Yes, modified.